substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/SubstitutionReportFeature.java

```diff
@@ -0,0 +1,188 @@
+            }
+        }
+    }
+
+    private void findSubstitutedFields(FeatureImpl.AfterAnalysisAccessImpl access) {
+        AnalysisUniverse universe = access.getUniverse();
+        for (AnalysisField field : universe.getFields()) {
+            if (field.wrapped instanceof SubstitutionField) {
+                SubstitutionField subField = (SubstitutionField) field.wrapped;
+                if (subField.isUserSubstitution()) {
+                    String jarLocation = getJarLocation(subField.getAnnotated().getDeclaringClass());
+                    substitutions.putIfAbsent(jarLocation, new Substitutions());
+                    substitutions.get(jarLocation).addField(subField);
+                }
+            }
+        }
+    }
+
+    private void reportSubstitutions() {
+        ReportUtils.report("substitutions performed by native-image","reports", "substitutions", "csv", pw -> {
+            pw.println("location , category (type/method/field) , original , annotated");
+            for (Map.Entry<String, Substitutions> g : substitutions.entrySet()) {
+                for (Map.Entry<ResolvedJavaType, ResolvedJavaType> e : g.getValue().getSubstitutedTypes().entrySet()) {
+                    pw.println(formatSubstitution(g.getKey(), "type", e.getKey(), e.getValue(), t -> t.toJavaName(true)));
+                }
+                for (Map.Entry<ResolvedJavaMethod, ResolvedJavaMethod> e : g.getValue().getSubstitutedMethods().entrySet()) {
+                    pw.println(formatSubstitution(g.getKey(), "method", e.getKey(), e.getValue(), this::formatMethod));
+                }
+                for (Map.Entry<ResolvedJavaField, ResolvedJavaField> e : g.getValue().getSubstitutedFields().entrySet()) {
+                    pw.println(formatSubstitution(g.getKey(), "field", e.getKey(), e.getValue(), this::formatField));
+                }
+            }
+        });
+    }
+
+    private String getJarLocation(ResolvedJavaType type) {
+        Class<?> annotatedClass = OriginalClassProvider.getJavaClass(GraalAccess.getOriginalSnippetReflection(), type);
+        CodeSource source = annotatedClass.getProtectionDomain().getCodeSource();
+        return source == null ? "unknown" : source.getLocation().toString();
+    }
+
+    private String formatMethod(ResolvedJavaMethod method) {
+        return method.getDeclaringClass().toJavaName(true) + "::" + method.getName();
+    }
+
+    private String formatField(ResolvedJavaField field) {
+        return field.getDeclaringClass().toJavaName(true) + "#" + field.getName();
+    }
+
+    private <T> String formatSubstitution(String jar, String type, T original, T annotated, Function<T, String> formatter) {
+        return '\'' + jar + "'," + type + ',' + formatter.apply(original) + ',' + formatter.apply(annotated);
+    }
+
+    private static class Substitutions {
+        private final Map<ResolvedJavaType, ResolvedJavaType> substitutedTypes = new TreeMap<>(new ResolvedJavaTypeComparator());
```

Comment:
Is there a reason to keep this sorted all the time as opposed to sorting before output?

Reply:
Not really, I simply preferred using a sorted collection, in order to keep the `reportSubstitutions` method simple. Do you want me to change that?

Reply:
Right, but we introduce all the comparators for that. Can we simplify the comparators somehow, e.g., by using a lambda?

Reply:
Yes. Changed.